<h2 style="text-align: center;"><b>Библиотека <a href="http://pandas.pydata.org/">pandas</a></b></h2>

Библиотека `pandas` активно используется в современном data science для работы с данными, которые могут быть представлены в виде таблиц (а это очень, очень большая часть данных)

`pandas` есть в пакете Anaconda, но если вдруг у Вас её по каким-то причинам нет, то можно установить, раскомментировав следующую команду:

In [0]:
#!pip install pandas

In [1]:
import pandas as pd
import numpy as np

Нужны данные. Про AppStore слышали? :)

### Скачать данные: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps

Нас интересует файл `AppStore.csv`.

Кстати, `.csv` (Comma Separated Values) - наверное, самый частый формат данных в современном data science. По сути, это то же самое, что и `.xls` (и `.xlsx`), то есть таблица. 

Итак, *чтение файла с данными*:

In [2]:
!wget 

"wget" ­Ґ пў«пҐвбп ў­гв७­Ґ© Ё«Ё ў­Ґи­Ґ©
Є®¬ ­¤®©, ЁбЇ®«­пҐ¬®© Їа®Ја ¬¬®© Ё«Ё Ї ЄҐв­л¬ д ©«®¬.


In [4]:
data = pd.read_csv('AppleStore.csv')

Посмотрим, что же такое переменная `data`:

In [5]:
data

Unnamed: 0          id  \
0              1   281656475   
1              2   281796108   
2              3   281940292   
3              4   282614216   
4              5   282935706   
...          ...         ...   
7192       11081  1187617475   
7193       11082  1187682390   
7194       11087  1187779532   
7195       11089  1187838770   
7196       11097  1188375727   

                                             track_name  size_bytes currency  \
0                                       PAC-MAN Premium   100788224      USD   
1                             Evernote - stay organized   158578688      USD   
2       WeatherBug - Local Weather, Radar, Maps, Alerts   100524032      USD   
3     eBay: Best App to Buy, Sell, Save! Online Shop...   128512000      USD   
4                                                 Bible    92774400      USD   
...                                                 ...         ...      ...   
7192                                              Kubik   126644224      USD   
7193                                  VR Roller-Coaster   120760320      USD   
7194              Bret Michaels Emojis + Lyric Keyboard   111322112      USD   
7195          VR Roller Coaster World - Virtual Reality    97235968      USD   
7196                       Escape the Sweet Shop Series    90898432      USD   

      price  rating_count_tot  rating_count_ver  user_rating  user_rating_ver  \
0      3.99             21292                26          4.0              4.5   
1      0.00            161065                26          4.0              3.5   
2      0.00            188583              2822          3.5              4.5   
3      0.00            262241               649          4.0              4.5   
4      0.00            985920              5320          4.5              5.0   
...     ...               ...               ...          ...              ...   
7192   0.00               142                75          4.5              4.5   
7193   0.00                30                30          4.5              4.5   
7194   1.99                15                 0          4.5              0.0   
7195   0.00                85                32          4.5              4.5   
7196   0.00                 3                 3          5.0              5.0   

         ver cont_rating   prime_genre  sup_devices.num  ipadSc_urls.num  \
0      6.3.5          4+         Games               38                5   
1      8.2.2          4+  Productivity               37                5   
2      5.0.0          4+       Weather               37                5   
3     5.10.0         12+      Shopping               37                5   
4      7.5.1          4+     Reference               37                5   
...      ...         ...           ...              ...              ...   
7192     1.3          4+         Games               38                5   
7193     0.9          4+         Games               38                0   
7194   1.0.2          9+     Utilities               37                1   
7195  1.0.15         12+         Games               38                0   
7196     1.0          4+         Games               40                0   

      lang.num  vpp_lic  
0           10        1  
1           23        1  
2            3        1  
3            9        1  
4           45        1  
...        ...      ...  
7192         1        1  
7193         1        1  
7194         1        1  
7195         2        1  
7196         2        1  

[7197 rows x 17 columns]

In [6]:
type(data)

pandas.core.frame.DataFrame

Тип == `pandas.core.frame.DataFrame`, обычно говорят просто "датафрейм", то есть "кусок данных".

Можно вывести информацию по датафрейму:

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7197 entries, 0 to 7196
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        7197 non-null   int64  
 1   id                7197 non-null   int64  
 2   track_name        7197 non-null   object 
 3   size_bytes        7197 non-null   int64  
 4   currency          7197 non-null   object 
 5   price             7197 non-null   float64
 6   rating_count_tot  7197 non-null   int64  
 7   rating_count_ver  7197 non-null   int64  
 8   user_rating       7197 non-null   float64
 9   user_rating_ver   7197 non-null   float64
 10  ver               7197 non-null   object 
 11  cont_rating       7197 non-null   object 
 12  prime_genre       7197 non-null   object 
 13  sup_devices.num   7197 non-null   int64  
 14  ipadSc_urls.num   7197 non-null   int64  
 15  lang.num          7197 non-null   int64  
 16  vpp_lic           7197 non-null   int64  


И более статистически значимую информацию:

In [8]:
data.describe()

Unnamed: 0            id    size_bytes        price  \
count   7197.000000  7.197000e+03  7.197000e+03  7197.000000   
mean    4759.069612  8.631310e+08  1.991345e+08     1.726218   
std     3093.625213  2.712368e+08  3.592069e+08     5.833006   
min        1.000000  2.816565e+08  5.898240e+05     0.000000   
25%     2090.000000  6.000937e+08  4.692275e+07     0.000000   
50%     4380.000000  9.781482e+08  9.715302e+07     0.000000   
75%     7223.000000  1.082310e+09  1.819249e+08     1.990000   
max    11097.000000  1.188376e+09  4.025970e+09   299.990000   

       rating_count_tot  rating_count_ver  user_rating  user_rating_ver  \
count      7.197000e+03       7197.000000  7197.000000      7197.000000   
mean       1.289291e+04        460.373906     3.526956         3.253578   
std        7.573941e+04       3920.455183     1.517948         1.809363   
min        0.000000e+00          0.000000     0.000000         0.000000   
25%        2.800000e+01          1.000000     3.500000         2.500000   
50%        3.000000e+02         23.000000     4.000000         4.000000   
75%        2.793000e+03        140.000000     4.500000         4.500000   
max        2.974676e+06     177050.000000     5.000000         5.000000   

       sup_devices.num  ipadSc_urls.num     lang.num      vpp_lic  
count      7197.000000      7197.000000  7197.000000  7197.000000  
mean         37.361817         3.707100     5.434903     0.993053  
std           3.737715         1.986005     7.919593     0.083066  
min           9.000000         0.000000     0.000000     0.000000  
25%          37.000000         3.000000     1.000000     1.000000  
50%          37.000000         5.000000     1.000000     1.000000  
75%          38.000000         5.000000     8.000000     1.000000  
max          47.000000         5.000000    75.000000     1.000000

Итак, наши данные -- это какая-то **информация про приложения в AppStore**. Давайте посмотрим, что тут есть:

In [9]:
# все столбцы датафрейма (названия)
data.columns

Index(['Unnamed: 0', 'id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')

---

#### Задание 0

О чём данные? (судя по столбцам)

**Ответ:** На самом деле у Kaggle-датасета всегда есть [описание на Kaggle](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) :)

---

Все значения датафрейма:

In [10]:
# одна строка == описание одного объекта, в данном случае -- приложения из AppStore
data.values

array([[1, 281656475, 'PAC-MAN Premium', ..., 5, 10, 1],
       [2, 281796108, 'Evernote - stay organized', ..., 5, 23, 1],
       [3, 281940292, 'WeatherBug - Local Weather, Radar, Maps, Alerts',
        ..., 5, 3, 1],
       ...,
       [11087, 1187779532, 'Bret Michaels Emojis + Lyric Keyboard', ...,
        1, 1, 1],
       [11089, 1187838770, 'VR Roller Coaster World - Virtual Reality',
        ..., 0, 2, 1],
       [11097, 1188375727, 'Escape the Sweet Shop Series', ..., 0, 2, 1]],
      dtype=object)

In [11]:
data.values[0]

array([1, 281656475, 'PAC-MAN Premium', 100788224, 'USD', 3.99, 21292, 26,
       4.0, 4.5, '6.3.5', '4+', 'Games', 38, 5, 10, 1], dtype=object)

In [12]:
data.shape

(7197, 17)

-- то есть 7197 строк и 17 столбцов.

Важно уметь обращаться к определённой строке или столбцу dataframe'а. Индексирование очень похоже на numpy, однако есть свои тонкости:

* Получить весь столбец (в даном случае столбец 'track_name'):

In [13]:
data['track_name']

0                                         PAC-MAN Premium
1                               Evernote - stay organized
2         WeatherBug - Local Weather, Radar, Maps, Alerts
3       eBay: Best App to Buy, Sell, Save! Online Shop...
4                                                   Bible
                              ...                        
7192                                                Kubik
7193                                    VR Roller-Coaster
7194                Bret Michaels Emojis + Lyric Keyboard
7195            VR Roller Coaster World - Virtual Reality
7196                         Escape the Sweet Shop Series
Name: track_name, Length: 7197, dtype: object

In [14]:
type(data['track_name'])

pandas.core.series.Series

In [15]:
data['track_name'][0]

'PAC-MAN Premium'

`pandas.core.series.Series` -- это тип подвыборки датафрейма. С этим типом работают реже, но стоит знать, что он есть.

* Получить определённую строку:

In [16]:
data.iloc[10]

Unnamed: 0                                 11
id                                  284791396
track_name          Solitaire by MobilityWare
size_bytes                           49618944
currency                                  USD
price                                    4.99
rating_count_tot                        76720
rating_count_ver                         4017
user_rating                               4.5
user_rating_ver                           4.5
ver                                    4.10.1
cont_rating                                4+
prime_genre                             Games
sup_devices.num                            38
ipadSc_urls.num                             4
lang.num                                   11
vpp_lic                                     1
Name: 10, dtype: object

Более продвинутое и полезное индексирование:

In [17]:
data.iloc[5, 10]

'1.8'

-- `data.iloc[i, j]`, где `i` -- номер строки, `j` -- номер столбца

Почти все типы индексирования:

* к строкам по числовому индексу, к столбцам -- по имени

In [18]:
data.loc[[0, 1, 2], ['track_name', 'id']]

track_name         id
0                                  PAC-MAN Premium  281656475
1                        Evernote - stay organized  281796108
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292

* к строкам и к столбцам по числовому индексу

In [19]:
data.iloc[[1, 3, 5], [2, 1]]

track_name         id
1                          Evernote - stay organized  281796108
3  eBay: Best App to Buy, Sell, Save! Online Shop...  282614216
5                                   Shanghai Mahjong  283619399

* поддерживает и то, и то, но **запрещён**:

In [22]:
data.ix[[2,1,0], ['track_name', 'id']]

AttributeError: 'DataFrame' object has no attribute 'ix'

In [23]:
data.ix[[0,1,2], [2, 1]]

AttributeError: 'DataFrame' object has no attribute 'ix'

Точно также дело обстоит со **срезами**:

In [24]:
data.iloc[5:100, 0:5]

Unnamed: 0         id                                   track_name  \
5            6  283619399                             Shanghai Mahjong   
6            7  283646709       PayPal - Send and request money safely   
7            8  284035177                      Pandora - Music & Radio   
8            9  284666222                  PCalc - The Best Calculator   
9           10  284736660                                  Ms. PAC-MAN   
..         ...        ...                                          ...   
95         108  306169895           Tozzle - Toddler's favorite puzzle   
96         109  306257910                         HomeBudget with Sync   
97         110  306375551                                   Wind Meter   
98         111  306468004  Map My Run+ - GPS Running & Workout Tracker   
99         112  306561234   Map My Ride+ - GPS Cycling & Route Tracker   

    size_bytes currency  
5     10485713      USD  
6    227795968      USD  
7    130242560      USD  
8     49250304      USD  
9     70023168      USD  
..         ...      ...  
95   168192000      USD  
96    17485824      USD  
97     5234688      USD  
98   237913088      USD  
99   233513984      USD  

[95 rows x 5 columns]

Но обычно нужно уметь отвечать на более осмысленные вопросы, например:

* сколько приложений имеют рейтинг юзеров >= 4?

In [25]:
len(data[data['user_rating'] >= 4])

4781

* какие есть валюты в столбце currency?

In [26]:
np.unique(data['currency'])

array(['USD'], dtype=object)

* сколько приложений жанра Games? А тех, которые Games и рейтинг >= 4?

In [27]:
# 1
len(data[data['prime_genre'] == 'Games'])

3862

In [28]:
# 2
len(data[(data['user_rating'] >= 4) & (data['prime_genre'] == 'Games')])

2868

Оператор "ИЛИ" аналогично &, только он через |

In [29]:
len(data[(data['user_rating'] >= 4) | (data['prime_genre'] == 'Games')])

5775

---

#### Задание 1

Получите ответ на следующие вопросы (с помощью pandas):

1). Сколько всего жанров и какие они?  
2). Сколько суммарно байт весят все приложения жанра Finance?  
3). Сколько всего бесплатных приложений, у которых рейтинг юзеров больше трёх?  
4). Есть ли приложения, которые стоят больше 100 долларов?

In [30]:
# 1
print(len(np.unique(data['prime_genre'])))
np.unique(data['prime_genre'])

23


array(['Book', 'Business', 'Catalogs', 'Education', 'Entertainment',
       'Finance', 'Food & Drink', 'Games', 'Health & Fitness',
       'Lifestyle', 'Medical', 'Music', 'Navigation', 'News',
       'Photo & Video', 'Productivity', 'Reference', 'Shopping',
       'Social Networking', 'Sports', 'Travel', 'Utilities', 'Weather'],
      dtype=object)

In [31]:
# 2
data[data['prime_genre'] == 'Finance']['size_bytes'].sum()

8136529366

In [32]:
# 3
len(data[(data['price'] == 0.0) & (data['user_rating'] > 3.0)])

2944

In [33]:
# 4
# по-хорошему надо ещё проверить, что валюта == 'USD', но мы уже поняли, что в этих данных только USD
len(data[data['price'] > 100])

2

---

Часто надо сделать что-то такое:

In [34]:
import numpy as np

In [35]:
np.array(data['size_bytes'])

array([100788224, 158578688, 100524032, ..., 111322112,  97235968,
        90898432], dtype=int64)

То есть взять столбец из таблицы и засунуть его в np.array(). Это можно сделать и просто обратившись к .values, ведь его тип и есть np.array:

In [36]:
type(data['size_bytes'].values)

numpy.ndarray

Больше про pandas можно найти по этом полезным ссылкам:

* Официальные туториалы: http://pandas.pydata.org/pandas-docs/stable/tutorials.html

* Статья на Хабре от OpenDataScience сообщества: https://habr.com/company/ods/blog/322626/

* Подробный гайд: https://media.readthedocs.org/pdf/pandasguide/latest/pandasguide.pdf

Главное в работе с новыми библиотеками -- не бояться тыкать в разные функции, смотреть типы возвращаемых объектов и активно пользоваться Яндексом, а ещё лучше понимать всё из docstring'а (`Shift+Tab` при нахождении курсора внутри скобок функции).